### DISCUSSION TITLE
Shape mismatch error when trying to get second order partial derivative of an integrator

### DISCUSSION THREAD

**Discussion by eaglema** [casadi/casadi#3281] [Q&A]
Hi everyone,

I'm trying to get second-order partial derivative $\frac{\partial^2{x_f}}{\partial{p}\partial{x_0}}$. But there is a shape mismatch error reported if I use adjoint sensitivity for the first-order derivative before calculating the second derivative, although the forward mode works fine. Here is a simple demo code to illustrate the problem.

import casadi as cas
x = cas.SX.sym('x', 3)
p = cas.SX.sym('p', 2)

ode1 = -p[0] * x[0]
ode2 = -p[1] * x[1]
ode3 = -p[1] * x[2]
ode_prob = {'x': x, 'p': p, 'ode': cas.vcat([ode1, ode2, ode3])}

#When ad_weight is 1, adjoint mode will be used
ode_fun = cas.integrator('ode_fun', 'cvodes', ode_prob, {'ad_weight': 1})

x0 = cas.MX.sym('x0', 3)
pm = cas.MX.sym('pm', 2)
sol = ode_fun(x0=x0, p=pm)
xf = sol['xf']

jac = cas.jacobian(xf, pm)
hes = cas.jacobian(jac, x0)

The errors are as followings:

RuntimeError: Error in MX::jacobian at .../casadi/core/mx.cpp:1708:
Error in XFunction::jac for 'helper_jacobian_MX' [MXFunction] at .../casadi/core/x_function.hpp:804:
Error in MXFunction::ad_forward at .../casadi/core/mx_function.cpp:890:
Error in MX::ad_forward for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2301:
Error in Call::ad_forward for 'jac_wrap_ode_fun' [MXFunction] at .../casadi/core/casadi_call.cpp:121:
Error in Function::forward for 'jac_wrap_ode_fun' [MXFunction] at .../casadi/core/function.cpp:1083:
Error in XFunction::get_forward for 'jac_wrap_ode_fun' [MXFunction] at .../casadi/core/x_function.hpp:856:
Error in MXFunction::ad_forward at .../casadi/core/mx_function.cpp:890:
Error in MX::ad_forward for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2301:
Error in Call::ad_forward for 'adj2_ode_fun' [MXFunction] at .../casadi/core/casadi_call.cpp:121:
Error in XFunction::get_jacobian for 'adj2_ode_fun' [MXFunction] at .../casadi/core/x_function.hpp:961:
Error in XFunction::jac for 'flattened_jac_adj2_ode_fun' [MXFunction] at .../casadi/core/x_function.hpp:804:
Error in MXFunction::ad_reverse at .../casadi/core/mx_function.cpp:1101:
Error in MX::ad_reverse for node of type N6casadi4CallE at .../casadi/core/mx.cpp:2310:
Error in Call::ad_reverse for 'asens2_ode_fun' [CvodesInterface] at .../casadi/core/casadi_call.cpp:149:
Error in Function::reverse for 'asens2_ode_fun' [CvodesInterface] at .../casadi/core/function.cpp:1091:
/casadi/core/sparsity_internal.cpp:2710: Assertion "numel() == nrow*ncol" failed:
reshape: number of elements must remain the same. Old shape is 39x1. New shape is 3x7=21.

**Comment by jgillis** [comment]
Thanks alot for sharing this regression bug! Spawned an issue

**Reply by eaglema** [reply]
Thanks, Joris! Look forward to the revised version!
